### PR TITLE
Add index_put, index_put_ to ATen XLA tensor

### PR DESCRIPTION
**Pull request by asuhan**

### CODE REVIEW

File: torch_xla/csrc/xla_lower_util.cpp [hunk 1]
```diff
@@ -247,4 +248,62 @@ xla::XlaOp CreateIndex(const xla::XlaOp& input, const xla::XlaOp& indices) {
...
+  }
+  for (xla::int64 i = 0; i < num_index_dims; ++i) {
+    dim_numbers.add_inserted_window_dims(i);
+    dim_numbers.add_scatter_dims_to_operand_dims(i);
+  }
+  // Build the combiner computation.
+  xla::XlaComputation combiner_computation;
+  {
+    xla::XlaBuilder cb("scatter-combiner");
+    auto xla_scalar_shape =
```
Comment (dlibenzi) [inline torch_xla/csrc/xla_lower_util.cpp]: xla::Shape

File: torch_xla/csrc/xla_lower_util.cpp [hunk 2]
```diff
@@ -247,4 +248,62 @@ xla::XlaOp CreateIndex(const xla::XlaOp& input, const xla::XlaOp& indices) {
...
+        xla::ShapeUtil::MakeShape(buffer_shape.element_type(), {});
+    auto p0 = xla::Parameter(&cb, 0, xla_scalar_shape, "p0");
```
Comment (dlibenzi) [inline torch_xla/csrc/xla_lower_util.cpp]: xla::XlaOp
(same below)

File: torch_xla/csrc/xla_lower_util.cpp [hunk 3]
```diff
@@ -247,4 +248,62 @@ xla::XlaOp CreateIndex(const xla::XlaOp& input, const xla::XlaOp& indices) {
...
+       i < values_rank; ++i) {
+    dim_numbers.add_update_window_dims(i);
+  }
+  for (xla::int64 i = 0; i < num_index_dims; ++i) {
+    dim_numbers.add_inserted_window_dims(i);
+    dim_numbers.add_scatter_dims_to_operand_dims(i);
+  }
+  // Build the combiner computation.
+  xla::XlaComputation combiner_computation;
+  {
```
Comment (dlibenzi) [inline torch_xla/csrc/xla_lower_util.cpp]: I don't think there is the need for this fake scope.

File: torch_xla/csrc/xla_lower_util.cpp [hunk 4]
```diff
@@ -247,4 +248,62 @@ xla::XlaOp CreateIndex(const xla::XlaOp& input, const xla::XlaOp& indices) {
   return xla::Gather(input, indices, dim_numbers, slice_sizes);
 }
 
+xla::XlaOp CreateIndexUpdate(
+    const xla::XlaOp& buffer, const xla::XlaOp& indices,
+    const xla::XlaOp& values,
+    const std::function<xla::XlaOp(xla::XlaOp, xla::XlaOp, xla::XlaBuilder*)>&
+        combiner) {
+  const xla::Shape buffer_shape = XlaHelpers::ShapeOfXlaOp(buffer);
```
Comment (dlibenzi) [inline torch_xla/csrc/xla_lower_util.cpp]: --const

File: torch_xla/csrc/ops/index_op.cpp [hunk 5]
```diff
@@ -127,6 +130,35 @@ ir::NodePtr IndexOp(const ir::Value& base, const ir::Value& indices) {
...
+  };
+  auto lower_for_shape_fn =
+      [&](tensorflow::gtl::ArraySlice<const xla::XlaOp> operands)
+      -> xla::XlaOp {
+    // The combiner doesn't matter for shape.
+    return CreateIndexUpdate(operands[0], operands[1], operands[2], nullptr);
+  };
+  return ir::ops::GenericOp(
+      ir::OpKind(at::aten::index_put), {buffer, indices, values},
+      ir::ops::InferOutputShape(
```
Comment (dlibenzi) [inline torch_xla/csrc/ops/index_op.cpp]: Use shape caching.

File: torch_xla/csrc/xla_lower_util.cpp [hunk 6]
```diff
@@ -247,4 +248,62 @@ xla::XlaOp CreateIndex(const xla::XlaOp& input, const xla::XlaOp& indices) {
...
+  const xla::Shape indices_shape = XlaHelpers::ShapeOfXlaOp(indices);
+  xla::Shape values_shape = XlaHelpers::ShapeOfXlaOp(values);
+
+  absl::Span<const xla::int64> indices_dims =
+      xla::AsInt64Slice(indices_shape.dimensions());
+  // The minor dimension of indices contains the indices to update.
+  xla::int64 num_index_dims = 1;
```
Comment (dlibenzi) [inline torch_xla/csrc/xla_lower_util.cpp]: Move below the CHECK.

File: torch_xla/csrc/xla_lower_util.cpp (hunk 1 above)
Comment (asuhan) [reply torch_xla/csrc/xla_lower_util.cpp]: Done.

File: torch_xla/csrc/xla_lower_util.cpp (hunk 2 above)
Comment (asuhan) [reply torch_xla/csrc/xla_lower_util.cpp]: Done.

File: torch_xla/csrc/xla_lower_util.cpp (hunk 3 above)
Comment (asuhan) [reply torch_xla/csrc/xla_lower_util.cpp]: Done.

File: torch_xla/csrc/xla_lower_util.cpp (hunk 4 above)
Comment (asuhan) [reply torch_xla/csrc/xla_lower_util.cpp]: Done.

File: torch_xla/csrc/ops/index_op.cpp (hunk 5 above)
Comment (asuhan) [reply torch_xla/csrc/ops/index_op.cpp]: Done.

File: torch_xla/csrc/xla_lower_util.cpp (hunk 6 above)
Comment (asuhan) [reply torch_xla/csrc/xla_lower_util.cpp]: Done.

**Pull request merged**